pkg/workflows/ring/store.go

```diff
@@ -0,0 +1,210 @@
+package ring
+
+import (
+	"context"
+	"maps"
+	"slices"
+	"sync"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/workflows/ring/pb"
+)
+
+// AllocationRequest represents a pending workflow allocation request during transition
+type AllocationRequest struct {
+	WorkflowID string
+	Result     chan uint32
+}
+
+// Store manages shard routing state and workflow mappings
+type Store struct {
+	routingState  map[string]uint32 // workflow_id -> shard_id (cache of allocated workflows)
+	shardHealth   map[uint32]bool   // shard_id -> is_healthy
+	healthyShards []uint32          // Sorted list of healthy shards
+	currentState  *pb.RoutingState  // Current routing state (steady or transition)
+
+	pendingAllocs map[string][]chan uint32 // workflow_id -> waiting channels
+	allocRequests chan AllocationRequest   // Channel for new allocation requests
+
+	mu sync.Mutex
+}
+
+const AllocationRequestChannelCapacity = 1000
+
+func NewStore() *Store {
+	return &Store{
+		routingState:  make(map[string]uint32),
+		shardHealth:   make(map[uint32]bool),
+		healthyShards: make([]uint32, 0),
+		pendingAllocs: make(map[string][]chan uint32),
+		allocRequests: make(chan AllocationRequest, AllocationRequestChannelCapacity),
+		mu:            sync.Mutex{},
+	}
+}
+
+func (s *Store) updateHealthyShards() {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	s.healthyShards = make([]uint32, 0)
+
+	for shardID, healthy := range s.shardHealth {
+		if healthy {
+			s.healthyShards = append(s.healthyShards, shardID)
+		}
+	}
+
+	// Sort for determinism
+	slices.Sort(s.healthyShards)
+
+	// If no healthy shards, add shard 0 as fallback
+	if len(s.healthyShards) == 0 {
+		s.healthyShards = []uint32{0}
+	}
+}
+
+func (s *Store) GetShardForWorkflow(ctx context.Context, workflowID string) (uint32, error) {
+	s.mu.Lock()
+
+	// Only trust the cache in steady state; during transition OCR may have invalidated it
+	if IsInSteadyState(s.currentState) {
+		// Check if already allocated in cache
+		if shard, ok := s.routingState[workflowID]; ok {
+			s.mu.Unlock()
+			return shard, nil
+		}
+		healthyShards := slices.Clone(s.healthyShards)
+		s.mu.Unlock()
+		return getShardForWorkflow(workflowID, healthyShards), nil
+	}
+
+	// During transition, defer to OCR consensus for consistent shard assignment across nodes
+	resultCh := make(chan uint32, 1)
+	s.pendingAllocs[workflowID] = append(s.pendingAllocs[workflowID], resultCh)
+	s.mu.Unlock()
+
+	select {
+	case s.allocRequests <- AllocationRequest{WorkflowID: workflowID, Result: resultCh}:
+	case <-ctx.Done():
+		return 0, ctx.Err()
+	}
+
+	select {
+	case shard := <-resultCh:
+		return shard, nil
+	case <-ctx.Done():
+		return 0, ctx.Err()
+	}
+}
+
+func (s *Store) SetShardForWorkflow(workflowID string, shardID uint32) {
```

Comment:
Add comment: This is called by the RingOCR plugin whenever it finishes a round with allocations for a given workflow ID.